internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_controller.go

```diff
@@ -480,3 +465,51 @@ func waitForUSBDelay() (err error) {
 
 	return nil
 }
+
+func (c *Controller) setRunning(ctx context.Context, seq runtime.Sequence) context.Context {
+	c.runningSequenceMu.Lock()
+	defer c.runningSequenceMu.Unlock()
+
+	if seq == -1 {
+		c.cancelCtx = nil
+		c.runningSequence = -1
+		c.Unlock()
+
+		return nil
+	}
+
+	ctx, c.cancelCtx = context.WithCancel(ctx)
+	c.runningSequence = seq
+
+	return ctx
+}
+
+func (c *Controller) tryTakeOver(ctx context.Context, seq runtime.Sequence, opts runtime.ControllerOptions) error {
+	return retry.Constant(time.Minute*1, retry.WithUnits(time.Millisecond*100)).RetryWithContext(ctx, func(ctx context.Context) error {
+		if c.cancelCtx != nil && c.runningSequence != -1 {
+			c.runningSequenceMu.Lock()
+			defer c.runningSequenceMu.Unlock()
+
+			canTakeOver := false
+
+			config, ok := sequenceTakeOver[c.runningSequence]
+			if ok {
+				_, canTakeOver = config[seq]
+			}
+
+			canTakeOver = canTakeOver || opts.Takeover
+
+			if !canTakeOver {
+				return runtime.ErrLocked
+			}
+
+			c.cancelCtx()
+		}
+
+		if c.TryLock() {
```

Comment:
we could probably have something simple as a lock, e.g. `chan` with buffer size of 1.

as long as there's a "token" there, you can grab it and run a sequence; once done, return token back to the pool

takeOver is all about cancelling the running sequence (if we are allowed to), and once cancelled, we grab the token to actually run the sequence

at the same time we want whoever cancelled the sequence (took it over) to run its sequence, so this probably becomes more tricky on this side.